src/spaces/discrete.jl

```diff
@@ -82,9 +82,24 @@ end
 Return a random position without any agents, or `nothing` if no such positions exist.
 """
 function random_empty(model::ABM{<:DiscreteSpace})
-    empty = empty_positions(model)
-    isempty(empty) && return nothing
-    rand(model.rng, collect(empty))
+    # This switch assumes the worst case (for this algorithm) of one
+    # agent per position, which is not true in general but is appropriate
+    # here.
+    if nagents(model) / prod(size(model.space)) < 0.998
```

Comment:
Perhaps we should put `0.998` as a second position argument? `random_empty(model, cutoff = 0.998)`? You never know, maybe in some weird other case, or higher dimensionalities, this needs to change.